### PR TITLE
feat: add `devcontainer.json`

### PR DESCRIPTION
**Pull request by okineadev**
As you know, in order to start working on this project, you need to install the necessary tools and packages, but this can take time.

For this purpose, Codespaces was created, so that you can **immediately** start development in the right environment, and not waste your time installing packages and configuring

And in order to implement it in your project - I added `devcontainer.json`

---

As you could understand, development in **Codespaces** will not start immediately, because for this you need to configure prebuilds (pre-installs the necessary packages and saves a system image ready for development)

After accepting the PR, I recommend enabling prebuilds with an update frequency of 4 times a week (apt will update packages 4 times a week to ensure fresh updates)

image

After that, developers can immediately start development in **Codespaces**

We can discuss what can be added to the configuration and what can be removed

### CODE REVIEW

**Review by okineadev — COMMENTED** [review]
knuckles-approved

**Review by Magrid0 — APPROVED** [review]
Looks good to me

**Comment by okineadev** [comment]
Why draft?

**Comment by kbdharun** [comment]
> Why draft?

I drafted it so that it doesn't get immediately merged as I need to review and discuss the changes. Will get back later today.

File: .devcontainer/devcontainer.json [hunk 1]
```diff
@@ -0,0 +1,32 @@
+// For format details, see https://aka.ms/devcontainer.json. For config options, see the
+// README at: https://github.com/devcontainers/templates/tree/main/src/ubuntu
+{
+  "name": "tldr-pages",
+  // Or use a Dockerfile or Docker Compose file. More info: https://containers.dev/guide/dockerfile
+  "image": "mcr.microsoft.com/devcontainers/base:jammy",
+  "features": {
```
Comment (okineadev) [inline .devcontainer/devcontainer.json:7]: **Python** and **NodeJS** are loaded in "features".

File: .devcontainer/devcontainer.json [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    "ghcr.io/devcontainers/features/node:1": {},
+    "ghcr.io/devcontainers/features/python:1": {}
```
Comment (okineadev) [inline .devcontainer/devcontainer.json:9]: Python and NodeJS are loaded in "features".

File: .devcontainer/devcontainer.json [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+  },
+
+  "privileged": true,
+
+  "onCreateCommand": {
+    "update-apt-packages": "sudo apt update && sudo apt upgrade -y",
```
Comment (okineadev) [inline .devcontainer/devcontainer.json:15]: This command updates apt-packages, this command takes some time, but with prebuilds you can save a few minutes (in prebuilds, all packages will be pre-installed)

File: .devcontainer/devcontainer.json [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    "install-python-packages": "pip install -r requirements.txt",
+    "install-node-packages": "npm install"
```
Comment (okineadev) [inline .devcontainer/devcontainer.json:17]: This is where the **Python** and **NodeJS** packages are installed

File: .devcontainer/devcontainer.json [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+  "customizations": {
+    "vscode": {
+      "settings": {
+        "resmon.show.battery": false,
+        "resmon.show.cpufreq": false
+      },
+      "extensions": [
+        "EditorConfig.EditorConfig",
+        "DavidAnson.vscode-markdownlint"
```
Comment (okineadev) [inline .devcontainer/devcontainer.json:28]: Here are the **vscode extensions** that will be installed on **Codespace**, I've added the [**EditorConfig**](https://marketplace.visualstudio.com/items?itemName=EditorConfig.EditorConfig) and [**Markdownlint**](https://marketplace.visualstudio.com/items?itemName=DavidAnson.vscode-markdownlint) extensions

If you know any extensions that will be useful when working in this repository - let me know

File: .devcontainer/devcontainer.json [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    "update-apt-packages": "sudo apt update && sudo apt upgrade -y",
+    "install-python-packages": "pip install -r requirements.txt",
+    "install-node-packages": "npm install"
+  },
+
+  "customizations": {
+    "vscode": {
+      "settings": {
+        "resmon.show.battery": false,
+        "resmon.show.cpufreq": false
```
Comment (okineadev) [inline .devcontainer/devcontainer.json:24]: Disabling CPU clock and battery indicators due to uselessness.

Here's where I got it from - https://github.com/microsoft/vscode/blob/main/.devcontainer/devcontainer.json#L25-L26

**Review by kbdharun — COMMENTED** [review]
Thanks for the explanations, I am currently testing the configuration locally and will get back in a few minutes. Having a devcontainer file is useful so that gets approval from me. Unfortunately, it isn't possible for us to setup pre-builds as it is a paid feature for organizations as mentioned in https://docs.github.com/en/codespaces/prebuilding-your-codespaces/configuring-prebuilds, we don't have a lot of dependencies so just having a dev container config would be enough as it would just take 30 seconds to 1 minute for it to get initialized fully.

**Review by kbdharun — CHANGES_REQUESTED** [review]
The current devcontainer configuration unfortunately takes a long time to build as you mentioned (locally it took me 15-20 minutes). ~~Since pre-builds aren't an option for us. A better alternative would be to tweak the Devcontainer configuration to use a pre-built Docker image instead along with these additional configs (this way the Codespace would take a relatively same time for initialization). I will mark this PR as a draft and will work on testing and porting this to it instead.~~

**Comment by okineadev** [comment]
> locally it took me 15-20 minutes

Then we can exclude "sudo apt upgrade -y"

---

> tweak the Devcontainer configuration to use a pre-built Docker image

Also, which prebuild Docker images are you talking about?

**Comment by kbdharun** [comment]
> Then we can exclude "sudo apt upgrade -y"

That would work too. Will test your config with this.

> Also, which prebuild Docker images are you talking about?

I meant building a Docker image with deps installed and specifying it in the dev-container config along with a custom command for our usecase.

**Comment by kbdharun** [comment]
> That would work too. Will test your config with this.

Yep, that worked perfectly which took the time to sub 1 minute for setup, I added a few extensions and made the container un-privileged (btw why was it set to privileged initially?).

**Review by kbdharun — APPROVED** [review]
LGTM, thanks for your contribution. _Tested this configuration with a codespace, the initialization took 1-2 minutes which IG is fine for most._

**Comment by okineadev** [comment]
Great, so you can merge the branch!

**Pull request merged**